### PR TITLE
In-memory cache improvements for template loading | TEC-3130

### PR DESCRIPTION
**Pull request by borkweb**
This PR adds some caching to frequently called operations that were degrading performance.

🎥 http://p.tri.be/Xpnrvd

:ticket: [TEC-3130]

[TEC-3130]: https://moderntribe.atlassian.net/browse/TEC-3130

### CODE REVIEW

File: src/Tribe/Main.php [hunk 1]
```diff
@@ -3807,17 +3807,25 @@ public function isEvent( $event ) {
 		 * @return bool Is it a venue?
 		 */
 		public function isVenue( $postId = null ) {
+			static $is_venue = [];
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: b82bcf4e0ce84f12e779538fb90f2661</sub>

File: src/Tribe/Main.php [hunk 2]
```diff
@@ -3807,17 +3807,25 @@ public function isEvent( $event ) {
...
+
 			if ( $postId === null || ! is_numeric( $postId ) ) {
 				global $post;
 				if ( isset( $post->ID ) ) {
 					$postId = $post->ID;
 				}
 			}
+
+			// Return if we've already fetched this info.
+			if ( isset( $is_venue[ $postId ] ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"

<sub>hash: 7113fd19f3781e24f128ad312befaff8</sub>

File: src/Tribe/Main.php [hunk 3]
```diff
@@ -3807,17 +3807,25 @@ public function isEvent( $event ) {
...
+				return $is_venue[ $postId ];
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"

<sub>hash: 03804ca5d877e2a31318a654544b18ca</sub>

File: src/Tribe/Main.php [hunk 4]
```diff
@@ -3807,17 +3807,25 @@ public function isEvent( $event ) {
...
+			}
+
 			if ( isset( $postId ) && get_post_field( 'post_type', $postId ) == Tribe__Events__Venue::POSTTYPE ) {
-				return true;
+				return $is_venue[ $postId ] = true;
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: * :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"
* :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: 478ceb50002f0cfd73f6b5dc65093b41</sub>

File: src/Tribe/Main.php [hunk 5]
```diff
@@ -3807,17 +3807,25 @@ public function isEvent( $event ) {
...
 			}
 
-			return false;
+			return $is_venue[ $postId ] = false;
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: * :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"
* :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: 94f19ca120be81bf9bec041805517e78</sub>

File: src/Tribe/Main.php [hunk 6]
```diff
@@ -3828,17 +3836,25 @@ public function isVenue( $postId = null ) {
 		 * @return bool Is it an organizer?
 		 */
 		public function isOrganizer( $postId = null ) {
+			static $is_organizer = [];
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: 59f9db1112aa8645f8604cd118acdcaf</sub>

File: src/Tribe/Main.php [hunk 7]
```diff
@@ -3828,17 +3836,25 @@ public function isVenue( $postId = null ) {
...
+
 			if ( $postId === null || ! is_numeric( $postId ) ) {
 				global $post;
 				if ( isset( $post->ID ) ) {
 					$postId = $post->ID;
 				}
 			}
+
+			// Return if we've already fetched this info.
+			if ( isset( $is_organizer[ $postId ] ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"

<sub>hash: 2585e5961c7ff464ae21167ee1c44079</sub>

File: src/Tribe/Main.php [hunk 8]
```diff
@@ -3828,17 +3836,25 @@ public function isVenue( $postId = null ) {
...
+				return $is_organizer[ $postId ];
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"

<sub>hash: b161023b24b3e3e69560074b6d328b8d</sub>

File: src/Tribe/Main.php [hunk 9]
```diff
@@ -3828,17 +3836,25 @@ public function isVenue( $postId = null ) {
...
+			}
+
 			if ( isset( $postId ) && get_post_field( 'post_type', $postId ) == Tribe__Events__Organizer::POSTTYPE ) {
-				return true;
+				return $is_organizer[ $postId ] = true;
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: * :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"
* :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: 3c214da1752716c0152cd96741d6cbc6</sub>

File: src/Tribe/Main.php [hunk 10]
```diff
@@ -3828,17 +3836,25 @@ public function isVenue( $postId = null ) {
...
 			}
 
-			return false;
+			return $is_organizer[ $postId ] = false;
```
Comment (tr1b0t) [inline src/Tribe/Main.php]: * :no_entry_sign: Variable "$postId" is not in valid snake_case format, try "$post_id"
* :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: 8760d926f63ebcd45d54c5a20e85edbf</sub>

File: src/Tribe/Timezones.php [hunk 11]
```diff
@@ -131,6 +131,14 @@ public static function event_end_timestamp( $event_id, $timezone = null ) {
 	 * @return int
 	 */
 	protected static function get_event_timestamp( $event_id, $type = 'Start', $timezone = null ) {
+		static $timestamps = [];
```
Comment (tr1b0t) [inline src/Tribe/Timezones.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: 9cfd502cfc3ac8032ed42b589e329e2e</sub>

File: src/Tribe/Timezones.php [hunk 12]
```diff
@@ -151,7 +159,7 @@ protected static function get_event_timestamp( $event_id, $type = 'Start', $time
 		if ( $use_event_tz || ( $use_site_tz && $site_zone_is_event_zone ) ) {
 			$datetime = get_post_meta( $event->ID, "_Event{$type}Date", true );
 
-			return strtotime( $datetime );
+			return $timestamps[ $cache_key ] = strtotime( $datetime );
```
Comment (tr1b0t) [inline src/Tribe/Timezones.php]: :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: 2f707d0188974c9c1690d90f2f27e7f0</sub>

File: src/Tribe/Timezones.php [hunk 13]
```diff
@@ -164,7 +172,7 @@ protected static function get_event_timestamp( $event_id, $type = 'Start', $time
 			: $timezone;
 
 		$localized = self::to_tz( $datetime, $tzstring );
-		return strtotime( $localized );
+		return $timestamps[ $cache_key ] = strtotime( $localized );
```
Comment (tr1b0t) [inline src/Tribe/Timezones.php]: :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: 9eee0aa67144a56b5d2de6c9d3e509c7</sub>

File: src/functions/template-tags/general.php [hunk 14]
```diff
@@ -1051,6 +1051,13 @@ function tribe_events_event_schedule_details( $event = null, $before = '', $afte
 			return '';
 		}
 
+		static $cache_details = [];
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: a959b4d632276a7288a7fbdddc39aabe</sub>

File: src/functions/template-tags/general.php [hunk 15]
```diff
@@ -1159,7 +1166,7 @@ function tribe_events_event_schedule_details( $event = null, $before = '', $afte
 		 * @param string $before    part of the HTML wrapper that was prepended
 		 * @param string $after     part of the HTML wrapper that was appended
 		 */
-		return apply_filters( 'tribe_events_event_schedule_details', $schedule, $event->ID, $before, $after );
+		return $cache_details[ $cache_details_key ] = apply_filters( 'tribe_events_event_schedule_details', $schedule, $event->ID, $before, $after );
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Assignments must be the first block of code on a line

<sub>hash: bdc7d12f9935a6fd58a35618f530b000</sub>

File: src/functions/template-tags/general.php [hunk 16]
```diff
@@ -1453,6 +1460,8 @@ function tribe_events_is_view_enabled( $view ) {
 	 * @return string|null Will return null on Bad Post Instances
 	 */
 	function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip_postdata_manipulation = false ) {
+		static $cache_excerpts = [];
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: 2157d11b0677256b21415cf8e076b796</sub>

File: src/functions/template-tags/general.php [hunk 17]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
 		 */
 		$remove_shortcodes = apply_filters( 'tribe_events_excerpt_shortcode_removal', true );
 
-		// Get the Excerpt or content based on what is available
-		$excerpt = has_excerpt( $post->ID ) ? $post->post_excerpt : $post->post_content;
+		$cache_excerpts_key = implode( ':', [
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: * :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call
* :no_entry_sign: Short array syntax is not allowed

<sub>hash: e2357dfba116fe974f6e887fb498e696</sub>

File: src/functions/template-tags/general.php [hunk 18]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+			$post->ID,
+			$skip_postdata_manipulation,
+			$allow_shortcodes,
+			$remove_shortcodes,
+			json_encode( $allowed_html )
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: * :warning: json_encode() is discouraged. Use wp_json_encode() instead.
* :no_entry_sign: Comma required after last value in array declaration
* :no_entry_sign: Each array item in a multi-line array declaration must end in a comma

<sub>hash: 861e1957da7830aba4c528766c1afeb7</sub>

File: src/functions/template-tags/general.php [hunk 19]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+		] );
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: 91e494b8bd3bc176e4511a8ded11f5bf</sub>

File: src/functions/template-tags/general.php [hunk 20]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+
+		if ( ! isset( $cache_excerpts[ $cache_excerpts_key ] ) ) {
+			// Get the Excerpt or content based on what is available
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 065d68b1a71f405216d036a480d7cff5</sub>

File: src/functions/template-tags/general.php [hunk 21]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+			$excerpt = has_excerpt( $post->ID ) ? $post->post_excerpt : $post->post_content;
+
+			// If shortcode filter is enabled let's process them
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: b61eb03058605f0e08ecca5e60dc4a1e</sub>

File: src/functions/template-tags/general.php [hunk 22]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+			if ( $allow_shortcodes ) {
+				$excerpt = do_shortcode( $excerpt );
+			}
 
-		// If shortcode filter is enabled let's process them
-		if ( $allow_shortcodes ) {
-			$excerpt = do_shortcode( $excerpt );
-		}
+			// Remove all shortcode Content before removing HTML
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 957934572ba945e34ac3448f36f3845e</sub>

File: src/functions/template-tags/general.php [hunk 23]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+			if ( $remove_shortcodes ) {
+				$excerpt = preg_replace( '#\[.+\]#U', '', $excerpt );
+			}
 
-		// Remove all shortcode Content before removing HTML
-		if ( $remove_shortcodes ) {
-			$excerpt = preg_replace( '#\[.+\]#U', '', $excerpt );
-		}
+			// Remove "all" HTML based on what is allowed
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 050fed7007e106a5f6857eefe4d53b96</sub>

File: src/functions/template-tags/general.php [hunk 24]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+			$excerpt = wp_kses( $excerpt, $allowed_html );
 
-		// Remove "all" HTML based on what is allowed
-		$excerpt = wp_kses( $excerpt, $allowed_html );
+			if ( ! has_excerpt( $post->ID ) ) {
+				// Temporarily alter the global post in preparation for our filters.
+				$global_post = isset( $GLOBALS['post'] ) ? $GLOBALS['post'] : null;
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :warning: Equals sign not aligned with surrounding assignments; expected 5 spaces but found 1 space

<sub>hash: 876af93f43bee4c3ad23d4917fe3c1a6</sub>

File: src/functions/template-tags/general.php [hunk 25]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+				$GLOBALS['post'] = $post;
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Overriding WordPress globals is prohibited. Found assignment to $GLOBALS['post']

<sub>hash: f0e745f0e21cdc29cec2d6a48de71d7d</sub>

File: src/functions/template-tags/general.php [hunk 26]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
 
-		if ( ! has_excerpt( $post->ID ) ) {
-			// Temporarily alter the global post in preparation for our filters.
-			$global_post = isset( $GLOBALS['post'] ) ? $GLOBALS['post'] : null;
-			$GLOBALS['post'] = $post;
+				// We will only trim Excerpt if it comes from Post Content
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: d0cd082985c0bf1701b08e3da6f95c11</sub>

File: src/functions/template-tags/general.php [hunk 27]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+				 */
+				$excerpt_more = apply_filters( 'excerpt_more', ' [&hellip;]' );
 
-			/**
-			 * Filter the string in the "more" link displayed after a trimmed excerpt.
-			 *
-			 * @param string $more_string The string shown within the more link.
-			 */
-			$excerpt_more = apply_filters( 'excerpt_more', ' [&hellip;]' );
+				// Now we actually trim it
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: b98f55bbd1ecc55994ea56c13000d202</sub>

File: src/functions/template-tags/general.php [hunk 28]
```diff
@@ -1518,48 +1527,60 @@ function tribe_events_get_the_excerpt( $post = null, $allowed_html = null, $skip
...
+				$excerpt = wp_trim_words( $excerpt, $excerpt_length, $excerpt_more );
 
-			// Now we actually trim it
-			$excerpt = wp_trim_words( $excerpt, $excerpt_length, $excerpt_more );
+				// Original post is back in action!
+				$GLOBALS['post'] = $global_post;
```
Comment (tr1b0t) [inline src/functions/template-tags/general.php]: :no_entry_sign: Overriding WordPress globals is prohibited. Found assignment to $GLOBALS['post']

<sub>hash: f9c7cb92ccc1f9cae4c5c606fb0322e6</sub>

File: src/views/v2/day/event.php [hunk 29]
```diff
@@ -16,7 +16,7 @@
  * @see tribe_get_event() For the format of the event object.
  */
 
-$classes = get_post_class( [ 'tribe-common-g-row', 'tribe-common-g-row--gutters', 'tribe-events-calendar-day__event' ], $event->ID );
+$classes = tribe_get_post_class( [ 'tribe-common-g-row', 'tribe-common-g-row--gutters', 'tribe-events-calendar-day__event' ], $event->ID );
```
Comment (tr1b0t) [inline src/views/v2/day/event.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: c4a6301f2b4ac3f35673fe1e3db913c9</sub>

File: src/views/v2/list/event.php [hunk 30]
```diff
@@ -19,7 +19,7 @@
 $container_classes = [ 'tribe-common-g-row', 'tribe-events-calendar-list__event-row' ];
 $container_classes['tribe-events-calendar-list__event-row--featured'] = $event->featured;
 
-$event_classes = get_post_class( [ 'tribe-events-calendar-list__event', 'tribe-common-g-row', 'tribe-common-g-row--gutters' ], $event->ID );
+$event_classes = tribe_get_post_class( [ 'tribe-events-calendar-list__event', 'tribe-common-g-row', 'tribe-common-g-row--gutters' ], $event->ID );
```
Comment (tr1b0t) [inline src/views/v2/list/event.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: e28191401aa05e13d077853ee88db6b6</sub>

File: src/views/v2/month/calendar-body/day/calendar-events/calendar-event.php [hunk 31]
```diff
@@ -16,7 +16,7 @@
  * @see tribe_get_event() For the format of the event object.
  */
 
-$classes = get_post_class( [ 'tribe-events-calendar-month__calendar-event' ], $event->ID );
+$classes = tribe_get_post_class( [ 'tribe-events-calendar-month__calendar-event' ], $event->ID );
```
Comment (tr1b0t) [inline src/views/v2/month/calendar-body/day/calendar-events/calendar-event.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: 974a393e0bd5bbb2e57aedbb21e1f9f3</sub>

File: src/views/v2/month/calendar-body/day/multiday-events/multiday-event.php [hunk 32]
```diff
@@ -31,7 +31,7 @@
 $should_display = $event->dates->start_display->format( 'Y-m-d' ) === $day_date
                   || $is_start_of_week;
 
-$classes = get_post_class( [ 'tribe-events-calendar-month__multiday-event' ], $event->ID );
+$classes = tribe_get_post_class( [ 'tribe-events-calendar-month__multiday-event' ], $event->ID );
```
Comment (tr1b0t) [inline src/views/v2/month/calendar-body/day/multiday-events/multiday-event.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: 44b4dad62a199198d5ebe4732fc99c56</sub>

File: src/views/v2/month/mobile-events/mobile-day/mobile-event.php [hunk 33]
```diff
@@ -15,7 +15,7 @@
  *
  * @see tribe_get_event() For the format of the event object.
  */
-$classes = get_post_class( [ 'tribe-events-calendar-month-mobile-events__mobile-event' ], $event->ID );
+$classes = tribe_get_post_class( [ 'tribe-events-calendar-month-mobile-events__mobile-event' ], $event->ID );
```
Comment (tr1b0t) [inline src/views/v2/month/mobile-events/mobile-day/mobile-event.php]: :no_entry_sign: Short array syntax is not allowed

<sub>hash: beb955b1047e66912c5093c4b6c266bc</sub>

File: src/Tribe/Main.php (hunk 1 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Lies.

File: src/Tribe/Main.php (hunk 2 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 3 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 4 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 5 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 6 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 7 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 8 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 9 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Main.php (hunk 10 above)
Comment (borkweb) [reply src/Tribe/Main.php]: Ignoring.

File: src/Tribe/Timezones.php (hunk 11 above)
Comment (borkweb) [reply src/Tribe/Timezones.php]: Ignoring.

File: src/Tribe/Timezones.php (hunk 12 above)
Comment (borkweb) [reply src/Tribe/Timezones.php]: Ignoring.

File: src/Tribe/Timezones.php (hunk 13 above)
Comment (borkweb) [reply src/Tribe/Timezones.php]: Ignoring.

File: src/views/v2/day/event.php (hunk 29 above)
Comment (borkweb) [reply src/views/v2/day/event.php]: Ignoring.

File: src/views/v2/list/event.php (hunk 30 above)
Comment (borkweb) [reply src/views/v2/list/event.php]: Ignoring.

File: src/views/v2/month/calendar-body/day/calendar-events/calendar-event.php (hunk 31 above)
Comment (borkweb) [reply src/views/v2/month/calendar-body/day/calendar-events/calendar-event.php]: Ignoring.

File: src/views/v2/month/calendar-body/day/multiday-events/multiday-event.php (hunk 32 above)
Comment (borkweb) [reply src/views/v2/month/calendar-body/day/multiday-events/multiday-event.php]: Ignoring.

File: src/views/v2/month/mobile-events/mobile-day/mobile-event.php (hunk 33 above)
Comment (borkweb) [reply src/views/v2/month/mobile-events/mobile-day/mobile-event.php]: Ignoring.

**Comment by bordoni** [comment]
Just tests failing.

**Pull request merged**